File: app/engine/ca_group.py
"""Utilities for computing case-age grouping buckets."""

from __future__ import annotations

from datetime import datetime


def get_ca_group(created_on: datetime) -> str:
    """Return the case-age group label based on the created timestamp."""

    now = datetime.now(tz=created_on.tzinfo)
    age_days = max((now - created_on).total_seconds() / 86_400, 0)

    if age_days <= 3:
        return "0–3 Days"
    if age_days <= 5:
        return "3–5 Days"
    if age_days <= 10:
        return "5–10 Days"
    if age_days <= 15:
        return "10–15 Days"
    if age_days <= 30:
        return "15–30 Days"
    if age_days <= 60:
        return "30–60 Days"
    if age_days <= 90:
        return "60–90 Days"
    return "> 90 Days"

File: src/engine/caGroup.js
const toValidDate = (value) => {
  if (value instanceof Date) {
    return Number.isNaN(value.getTime()) ? null : value;
  }
  const date = new Date(value);
  return Number.isNaN(date.getTime()) ? null : date;
};

const MS_PER_DAY = 1000 * 60 * 60 * 24;

export const getCaGroup = (createdOn) => {
  const createdDate = toValidDate(createdOn);
  if (!createdDate) {
    return "";
  }

  const now = new Date();
  const diffDays = (now - createdDate) / MS_PER_DAY;

  if (diffDays <= 3) {
    return "0–3 Days";
  }
  if (diffDays <= 5) {
    return "3–5 Days";
  }
  if (diffDays <= 10) {
    return "5–10 Days";
  }
  if (diffDays <= 15) {
    return "10–15 Days";
  }
  if (diffDays <= 30) {
    return "15–30 Days";
  }
  if (diffDays <= 60) {
    return "30–60 Days";
  }
  if (diffDays <= 90) {
    return "60–90 Days";
  }
  return "> 90 Days";
};
